### PR TITLE
Decouple GA population from frontend

### PR DESCRIPTION
**Pull request by MichaelAEden**
- Decouple GA population parameter from the number of bots displayed in the frontend
- Move GA configuration into an object (will prevent bugs when running experiments due to tests running in parallel)
- Change `/createbots/rating` to `/bots`
- Linting

### CODE REVIEW

File: server/src/ga/GaWorker.ts [hunk 1]
```diff
@@ -11,18 +18,32 @@ import { OctaveRule } from "../rules/OctaveRule";
...
-  static ITERATIONS = 5; // Times GA will iterate
-  // TODO decouple this
-  POPULATION_SIZE = 7; // Population size
-  static MUTATION_RATE = 0.15; // Probability of mutation
-  static NO_FAVOURITE_RATE = 0.5; // weight of selection is weaker if bot not favourited
-  static MUSICAL_FITNESS_WEIGHT = 1; // relative weight of melody vs user fitness in convention algo
+interface GaWorkerConfig {
+  iterations: number;
+  mutationRate: number;
+  noFavourateRate: number;
```
Comment (MickeyDang) [inline server/src/ga/GaWorker.ts]: `noFavouriteWeight` I think would be more descriptive of what it does.

File: server/src/ga/GaWorker.ts [hunk 2]
```diff
@@ -11,18 +18,34 @@ import { OctaveRule } from "../rules/OctaveRule";
...
-  static ITERATIONS = 5; // Times GA will iterate
-  // TODO decouple this
-  POPULATION_SIZE = 7; // Population size
-  static MUTATION_RATE = 0.15; // Probability of mutation
-  static NO_FAVOURITE_RATE = 0.5; // weight of selection is weaker if bot not favourited
-  static MUSICAL_FITNESS_WEIGHT = 1; // relative weight of melody vs user fitness in convention algo
+interface GaWorkerConfig {
+  iterations: number; // Times GA will iterate
+  mutationRate: number; // Probability of mutation
+  noFavourateRate: number; // Weight of selection is weaker if bot not favourited
```
Comment (MickeyDang) [inline server/src/ga/GaWorker.ts:24]: `noFavouriteWeight` would be more descriptive i think

File: server/src/ga/GaWorker.ts [hunk 3]
```diff
@@ -31,55 +54,46 @@ export class GaWorker {
...
         "G4,B4,D5,E5",
         "E4,E4,E4,A4",
         "B4,A4,G4,D4",
         "D4,G4,C5,D5",
-        "E4,G4,E4,D4"
-      ].map((str) => new Bot(0, Melody.fromString(str)));
-
+        "E4,G4,E4,D4",
+      ];
+      const selectedMelodies = selectRandomMany(melodies, this.config.selectionSize);
```
Comment (MickeyDang) [inline server/src/ga/GaWorker.ts:81]: yess i was hoping to implement this i think its much cleaner!

File: server/src/ga/GaWorker.ts [hunk 4]
```diff
@@ -31,55 +54,46 @@ export class GaWorker {
...
-      // Feature flagging
-
+    for (let i = 0; i < this.config.iterations; i++) {
       // Produce fitness scores from bots
-      let fitnesses = evaluate(generation);
+      fitnesses = evaluate(generation);
 
       // Normalize the scores to select a new generation
-      generation = selectRandomWeighted(generation, fitnesses, this.POPULATION_SIZE);
+      generation = selectRandomWeighted(generation, fitnesses, this.config.populationSize);
```
Comment (MickeyDang) [inline server/src/ga/GaWorker.ts:96]: Are we adding tests to make sure random weighted does what we want?
Also would be good to maybe have some hypothesis that this is actually what we want. Something related fitness score similar to some of our existing ones.

Comment (MichaelAEden) [reply server/src/ga/GaWorker.ts:96]: Already have tests for it! Check out `Utils.test.ts`

Comment (MickeyDang) [reply server/src/ga/GaWorker.ts:100]: sweet!

File: server/src/ga/FitnessConvention.ts [hunk 5]
```diff
@@ -62,17 +61,20 @@ export const evaluateWhoop = (melody: Melody): number => {
   return whoopFitness;
 };
 
-export const getUserFitness = (bot: Bot) => {
-  return bot.metric != 0 ? bot.metric : GaWorker.NO_FAVOURITE_RATE;
-}
+export const getUserFitness = (bot: Bot, noFavouriteRate: number) => {
```
Comment (MickeyDang) [inline server/src/ga/FitnessConvention.ts:64]: `noFavouriteWeight`

File: server/src/ga/FitnessConvention.ts [hunk 6]
```diff
@@ -62,17 +61,20 @@ export const evaluateWhoop = (melody: Melody): number => {
...
-  return (evaluateRange(bot.melody) + evaluateStepwise(bot.melody) + evaluateWhoop(bot.melody)) / MAX_MELODY_SCORE;
-}
+  return (
+    (evaluateRange(bot.melody) + evaluateStepwise(bot.melody) + evaluateWhoop(bot.melody)) /
+    MAX_MELODY_SCORE
+  );
+};
 
-export default (bots: Bot[]): number[] => {
+export default (bots: Bot[], noFavouriteRate: number, musicalFitnessWeight: number): number[] => {
```
Comment (MickeyDang) [inline server/src/ga/FitnessConvention.ts:76]: `noFavouriteWeight`

**Pull request merged**